rfd/xxxx-command.md

```diff
@@ -0,0 +1,208 @@
+users, then one command should be execute for each user.
+An alternative would be to add the user to `ssh_service` configuration, but this
+will require modifying the configuration on all nodes. 
+
+## Roles
+
+Roles will be extended to allow defining access to commands. The following
+example will allow to execute commands with `environment=dev-cmd` label on nodes
+with `environment=dev` label. It will also deny access to commands with
+`environment=prod` label.
+This approach will allow to define a set of commands that can be executed
+on a particular set of nodes (e.g. `dev-cmd` commands can be executed on `dev` nodes
+but not on `prod` nodes).
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: developer
+spec:
+  allow:
+    # Labels selector for command execution.
+    command_execution_labels:
+      environment: ["dev-cmd"]
+    # Labels selector for commands resource access.
+    command_labels:
+     environment: ["dev"]
+  deny:
+    command_labels:
+     environment: ["prod"]
+```
+
+Important thing to note is that a user could have access to a command but not
+the node itself (interactive ssh access). This will allow to harden the access
+to the node by allowing to execute only a specific set of commands.
+
+## tsh extensions
+
+`tsh` will learn new commands:
+
+1. `tsh command ls` - allowing to list all available commands.
+2. `tsh command exec` - allowing to execute commands on one or multiple nodes. User will be reqired to provide:
+   * node query using [predicate launguage](https://goteleport.com/docs/reference/predicate-language/) 
+   * OS user that will be used to execute the command
+3. `tsh command cancel` - (optional) allowing to cancel a command execution.
+4. `tsh command logs` - (optional) allowing to display output from a command execution.
+
+### Display output from multiple nodes
+
+Displaying the output from multiple nodes can be tricky as the output can be
+very long. To avoid this, Teleport Command will detect long output and will
+redirect it to a file. The user will be notified about this.
+
+```shell
+$ tsh command exec cat-log
+node1: Detected long output. Redirecting to node1.log file
+node2: Detected long output. Redirecting to node2.log file
+```
+
+Alternative approach is to always display the output if a command is executed
+on a single node and redirect it to a file if a command is executed on multiple nodes.
+
+### Per-session MFA
+
+By design per-session MFA requires "a tap" for every node where a command is executed.
+This is not ideal from the UX perspective as it will require a user to tap multiple times. 
+To avoid this, Teleport will ask for "a tap" only once before executing all commands.
+In the background, Teleport Proxy will generate certificates with mutliple principals
+for every node and will use them to authenticate to the nodes. If one certificate cannot
+carry all principals, multiple certificates will be generated.
+
```

Comment:
Missing a couple of extra sections as well:

- IaC. How will we integrate this with our terraform provider and Kube Operator?
- Teleport Connect and Web UI.